### DISCUSSION TITLE
VM + Managed Identity + KV

### DISCUSSION THREAD

**Discussion by szympulka** [AzureCommunityPL/group#434] [Pomoc] [answered]
Mam terraforma, który tworzy mi VM. Podczas tworzenia tej VM odpalam custom script, który wywołuje metodę `az keyvault secret show --id `. Konsola krzyczy, że nie mogę się dobić do KV ponieważ nie wykonałem Az Login. Próbowałem całość spiąć przez `system assigned managed identity` oraz `User assigned` - skrypt dalej woła aby zalogować się do KV. W jaki sposób mogę się zalogować/mieć dostęp do KV używając Managed Identity?
```
resource "azurerm_virtual_machine_extension" "setup_env" {
  name                 = "SetupEnv"
  virtual_machine_id   = azurerm_linux_virtual_machine.vm.id
  type                 = "CustomScript"
  type_handler_version = "2.0"
  publisher            = "Microsoft.Azure.Extensions"

  protected_settings = <<PROTECTED_SETTINGS
{
      "script": "${base64encode(file("setup.sh"))}"
}
PROTECTED_SETTINGS

  depends_on = [
    azurerm_linux_virtual_machine.vm
  ]
}

```

**Comment by kamil-mrzyglod** [comment, marked as answer]
Ogólnie bardziej z mojej strony "na intuicję" bo z tym mało się bawiłem.

1. Nawet jesli masz MI włączone na poziomie VM to tak magicznie Azure CLI sie nie połapie. Pytanie czy w ogóle wywołujesz "az login" a jeśli tak - czy z użyciem "--identity"
2. Pytanie czy nie możesz wywołać w ramach "script" dwóch poleceń na raz - "az login && setup.sh"
3. Protected_settings mają parametr "managedIdentity" - normalnie wykorzystywany do pobierania plików zamiast wykorzystywać Storage Account, może przy okazji zadziała na sam skrypt
4. Jak już masz TF to możesz użyć provisionera jako alternatywy zamiast custom script extension

Realnie musiałbym sobie sam to potestować :)

**Reply by szympulka** [reply]
1. Dokładnie tak. używam az login --identity :)
2. hmmm, zobaczmy czy łyknie w takim razie 
3. Sprawdzę
4. Możesz rozwinać?

**Reply by szympulka** [reply]
Nakierowałeś mnie Kamil by zagłębić się w az login i rozwiązaniem było `az login --identity --allow-no-subscriptions`

**Reply by kamil-mrzyglod** [reply]
Co do rozwinięcia, tutaj masz docsy - https://www.terraform.io/language/resources/provisioners/remote-exec

Odnośnie samego rozwiązania (tak trochę dla potomnych) - generanie `--allow-no-subscriptions` jest używane jeśli chce się zalogować do Azure w tenancie gdzie nie żadnych subskrypcji (albo do żadnej nie mamy dostępu) i chce się za pomocą Azure CLI operować na AAD (np. z pomocą `az ad`). To, że akurat tutaj zadziałało mnie trochę zaskoczyło, ale możliwe że:
- Identity nie jest przypięte na poziomie RBAC do żadnego z zasobów (bo nie musi) i bez `--allow-no-subscriptions` zwyczajnie nie puszcza logowania
- Pobieranie sekretów idzie po data plane więc nie trzeba tego wywoływać w kontekście jakiejkolwiek subskrypcji

Takie małe przypuszczenia :)
